File: Bubble_Sort/Objective-C/BubbleSort.m
//
//  BubbleSort.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "BubbleSort.h"

@implementation BubbleSort

+ (void)sort:(NSMutableArray<NSNumber *> *)array order:(BubbleSortOrder)order {
    NSUInteger arrayCount = array.count;

    if (arrayCount < 2) {
        return;
    }

    for (NSUInteger i = arrayCount; i > 0; i--) {
        for (NSUInteger j = 0; j < i - 1; j++) {
            NSNumber *currentNumber = array[j];
            NSUInteger nextIndex = j + 1;
            NSNumber *nextNumber = array[nextIndex];

            NSComparisonResult comparisonResult = [currentNumber compare:nextNumber];

            if ((order == BubbleSortOrderAsc && comparisonResult == NSOrderedDescending) ||
                (order == BubbleSortOrderDesc && comparisonResult == NSOrderedAscending)) {
                [array exchangeObjectAtIndex:j withObjectAtIndex:nextIndex];
            }
        }
    }
}

@end

File: Bubble_Sort/Swift/BubbleSort.swift
//
//  BubbleSort.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct BubbleSort {

    enum SortOrder {
        case ASC
        case DESC
    }

    /** Complexity is O(n^2) */
    static public func sort<T: Comparable>(_ array: [T?]?, order: SortOrder) -> [T?]? {
        let arrayCount = array?.count ?? 0

        if arrayCount < 2 {
            return array
        }

        var resultArray = array!

        for i in stride(from: arrayCount, to: 0, by: -1) {
            for j in 0..<(i - 1) {
                let currentValue = resultArray[j]
                let nextIndex = j + 1
                let nextValue = resultArray[nextIndex]

                if (order == .ASC && ((currentValue != nil && nextValue == nil) || (currentValue != nil && nextValue != nil && currentValue! > nextValue!))) ||
                    (order == .DESC && ((currentValue == nil && nextValue != nil) || (currentValue != nil && nextValue != nil && currentValue! < nextValue!))) {
                    swap(&resultArray[j], &resultArray[nextIndex])
                }
            }
        }

        return resultArray;
    }

}
